## ``Classification Models - Video Games``

##### ``Sandra Ranilla-Cortina``

In [19]:
# Libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

plt.style.use('fivethirtyeight')
np.random.seed(777)

import warnings
warnings.filterwarnings("ignore", category = UserWarning)
warnings.filterwarnings("ignore", category = FutureWarning)

#### ``Video Games Pre-Processed Data + Dimensionality Reduction (PCA)``

In [20]:
# Video Games Pre-Processed data + PCA
X_train = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/VideoGames/X_Train_PCA_VideoGames.pkl')
X_test = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/VideoGames/X_Test_PCA_VideoGames.pkl')
Y_train = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/VideoGames/Y_Train_VideoGames.pkl')
Y_test = pd.read_pickle('https://raw.githubusercontent.com/sranillac/SRC_UniOvi/refs/heads/main/ML-IA/data/VideoGames/Y_Test_VideoGames.pkl')
print(X_train.shape, X_test.shape)
print(Y_train.shape, Y_test.shape)

(35830, 6) (11944, 6)
(35830, 1) (11944, 1)


#### ``Logistic Regression``

In [21]:
# Logistic Regression Model (LR)
log_reg = LogisticRegression(max_iter = 1000, random_state = 42)
log_reg.fit(X_train, Y_train)

LogisticRegression(max_iter=1000, random_state=42)

In [22]:
# Prediction
Y_pred_LR = log_reg.predict(X_test)

#### ``Support Vector Machine (SVM)``

In [37]:
# Support Vector Machine Model (SVM)
svm_model = SVC(kernel = 'rbf', random_state = 42)
svm_model.fit(X_train, Y_train)

SVC(random_state=42)

In [38]:
# Prediction
Y_pred_SVM = svm_model.predict(X_test)

#### ``K-Neighbors``

In [25]:
# K-Neighors Model (KNN)
knn_model = KNeighborsClassifier(n_neighbors = 4)
knn_model.fit(X_train, Y_train)

KNeighborsClassifier(n_neighbors=4)

In [26]:
# Prediction
Y_pred_KNN = knn_model.predict(X_test)

#### ``Decision Tree``

In [27]:
# Decision Tree Model (DT)
dt_model = DecisionTreeClassifier(max_depth = 5, random_state = 42)
dt_model.fit(X_train, Y_train)

DecisionTreeClassifier(max_depth=5, random_state=42)

In [28]:
# Prediction
Y_pred_DT = dt_model.predict(X_test)

#### ``Random Forest``

In [29]:
# Random Forest Model (RF)
rf_model = RandomForestClassifier(n_estimators = 100, max_depth = 5, random_state = 42)
rf_model.fit(X_train, Y_train)

RandomForestClassifier(max_depth=5, random_state=42)

In [30]:
# Evaluation
Y_pred_RF = rf_model.predict(X_test)

#### ``Evaluation Metrics``

1) ``Confusion matrix`` is a table that allows us to visualize the performance of a classification model:  
    - True Positives (**TP**): number of instances correctly predicted as positive
    - True Negatives (**TN**): number of instances correctly predicted as negative
    - False Positives (**FP**): number of instances incorrectly predicted as positive
    - False Negatives (**FN**): number of instances incorrectly predicted as negative.  


2) ``Accuracy`` measures the proportion of correct predictions (both positive and negative) out of the total predictions made:  

    $\text{Accuracy} = \frac{TP + TN}{TP + TN + FP + FN}$  

3) ``Precision`` measures the proportion of positive predictions that are actually correct:  

    $\text{Precision} = \frac{TP}{TP + FP}$  

4) ``Recall`` measures the proportion of actual positive instances that were correctly identified by the model:  

     $\text{Recall} = \frac{TP}{TP + FN}$  

5) ``F1-Score`` is the harmonic mean between **Precision** and **Recall**. It is useful when there is an imbalance in the classes:

     $\text{F1-Score} = 2 \times \frac{\text{Precision} \times \text{Recall}}{\text{Precision} + \text{Recall}}$

In [31]:
# Evaluation Metrics function
def eval_metrics(y, y_pred):
    mc = confusion_matrix(y, y_pred)
    acc = accuracy_score(y, y_pred)
    cr = classification_report(y, y_pred)
    return mc, acc, cr

In [32]:
# Logistic Regression Model - Evaluation
mc_LR, acc_LR, cr_LR = eval_metrics(Y_test, Y_pred_LR)

print(f'Logistic Regression Model - Confusion Matrix:')
print(mc_LR)
print('')
print(f'Logistic Regression Model - Accuracy: {acc_LR:.2f}')
print('')
print(f'Logistic Regression Model - Classification Report:')
print(cr_LR)

Logistic Regression Model - Confusion Matrix:
[[  56 1357    0]
 [  68 9315   13]
 [   0 1130    5]]

Logistic Regression Model - Accuracy: 0.78

Logistic Regression Model - Classification Report:
              precision    recall  f1-score   support

           0       0.45      0.04      0.07      1413
           1       0.79      0.99      0.88      9396
           2       0.28      0.00      0.01      1135

    accuracy                           0.78     11944
   macro avg       0.51      0.35      0.32     11944
weighted avg       0.70      0.78      0.70     11944



In [39]:
# SVM Model - Evaluation
mc_SVM, acc_SVM, cr_SVM = eval_metrics(Y_test, Y_pred_SVM)

print(f'SVM Model - Confusion Matrix:')
print(mc_SVM)
print('')
print(f'SVM Model - Accuracy: {acc_SVM:.2f}')
print('')
print(f'SVM Model - Classification Report:')
print(cr_SVM)

SVM Model - Confusion Matrix:
[[   0 1413    0]
 [   0 9396    0]
 [   0 1135    0]]

SVM Model - Accuracy: 0.79

SVM Model - Classification Report:
              precision    recall  f1-score   support

           0       0.00      0.00      0.00      1413
           1       0.79      1.00      0.88      9396
           2       0.00      0.00      0.00      1135

    accuracy                           0.79     11944
   macro avg       0.26      0.33      0.29     11944
weighted avg       0.62      0.79      0.69     11944



In [34]:
# K-Neighbors Model - Evaluation
mc_KNN, acc_KNN, cr_KNN = eval_metrics(Y_test, Y_pred_KNN)

print(f'KNN Model - Confusion Matrix:')
print(mc_KNN)
print('')
print(f'KNN Model - Accuracy: {acc_KNN:.2f}')
print('')
print(f'KNN Model - Classification Report:')
print(cr_KNN)

KNN Model - Confusion Matrix:
[[ 440  973    0]
 [ 909 8316  171]
 [   0 1050   85]]

KNN Model - Accuracy: 0.74

KNN Model - Classification Report:
              precision    recall  f1-score   support

           0       0.33      0.31      0.32      1413
           1       0.80      0.89      0.84      9396
           2       0.33      0.07      0.12      1135

    accuracy                           0.74     11944
   macro avg       0.49      0.42      0.43     11944
weighted avg       0.70      0.74      0.71     11944



In [35]:
# Decision Tree Model - Evaluation
mc_DT, acc_DT, cr_DT = eval_metrics(Y_test, Y_pred_DT)

print(f'Decision Tree Model - Confusion Matrix:')
print(mc_DT)
print('')
print(f'Decision Tree Model - Accuracy: {acc_DT:.2f}')
print('')
print(f'Decision Tree Model - Classification Report:')
print(cr_DT)

Decision Tree Model - Confusion Matrix:
[[  53 1360    0]
 [  88 9301    7]
 [   0 1132    3]]

Decision Tree Model - Accuracy: 0.78

Decision Tree Model - Classification Report:
              precision    recall  f1-score   support

           0       0.38      0.04      0.07      1413
           1       0.79      0.99      0.88      9396
           2       0.30      0.00      0.01      1135

    accuracy                           0.78     11944
   macro avg       0.49      0.34      0.32     11944
weighted avg       0.69      0.78      0.70     11944



In [36]:
# Random Forest Model - Evaluation
mc_RF, acc_RF, cr_RF = eval_metrics(Y_test, Y_pred_RF)

print(f'Random Forest Model - Confusion Matrix:')
print(mc_RF)
print('')
print(f'Random Forest Model - Accuracy: {acc_RF:.2f}')
print('')
print(f'Random Forest Model - Classification Report:')
print(cr_RF)

Random Forest Model - Confusion Matrix:
[[   1 1412    0]
 [   2 9394    0]
 [   0 1135    0]]

Random Forest Model - Accuracy: 0.79

Random Forest Model - Classification Report:
              precision    recall  f1-score   support

           0       0.33      0.00      0.00      1413
           1       0.79      1.00      0.88      9396
           2       0.00      0.00      0.00      1135

    accuracy                           0.79     11944
   macro avg       0.37      0.33      0.29     11944
weighted avg       0.66      0.79      0.69     11944



Which is the **``best performing model``**?  

- The **Random Forest** model achieves $50\%$ accuracy, with decent recall for **class 1** and **class 2** ($0.63$). However, it struggles with **class 0** and **class 3**.  

- The **Decision Tree** model also has $50\%$ accuracy, but performs slightly worse for **class 0** and **class 3** compared to **Random Forest**.  

- The **KNN** model has the lowest accuracy ($43\%$) and performs poorly for **class 3**.

- The **SVM**, and **Logistic Regression** models both achieve $50\%$ accuracy, but struggle significantly with **class 0** and **class 3**. 

Therefore, **``Random Forest``** is the best model despite its limitations, offering the highest recall for **class 1** and **class 2** (maybe a re-balance / distribution of the target ``User Rating`` variable should be considered).